### ISSUE TITLE
Merge Issue in team subscription

### ISSUE THREAD

**Issue by infocodiste** [FlutterFlow/flutterflow-issues#2141]
### Has your issue been reported?

- [X] I have searched the existing issues and confirm it has not been reported.
- [X] I give permission for members of the FlutterFlow team to access and test my project for the sole purpose of investigating this issue.

### Current Behavior

It seems like I'm encountering a problem while merging branches in FlutterFlow. When I merge another branch into the main one, it ends up overwriting all the changes from the other branch, and the main branch loses its modifications. Additionally, attempting to rebase with the main branch causes the other branch to adopt the characteristics of the main branch and loses its modifications. This is quite frustrating for me as a developer, especially after investing in the enterprise-level access. After spending two days carefully merging two branches, I'm unable to find a well-merged branch, forcing me to resort to manual project management by manually write within the main branch

### Expected Behavior

When merging branches in FlutterFlow, the expected behavior is that the changes from both branches should be successfully combined into the main branch without overwriting each other. The merge process should intelligently handle conflicts, if any, allowing for a seamless integration of modifications from multiple branches. Additionally, when rebasing with the main branch, the other branch should retain its unique modifications while incorporating changes from the main branch, creating a harmonious and consolidated codebase. The process should not result in the loss of modifications or require manual intervention for proper project management

### Steps to Reproduce

Step to Reproduce issue with respect my project

1. Open new branch "create_event_new"
2. Click on Merge icon
3. Click on Merge \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\[create_event_new\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\] into \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\[main\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\]
4. Wait a few time, and observe, you can see there no conflict and you can merge without any issue.
5. Now move to main branch, you can see there code overwrite by new branch and your changes flush.

### Reproducible from Blank

- [X] The steps to reproduce above start from a blank project.

### Bug Report Code (Required)

IT40pvrf4oh/jOlm5ZDqePRWqBo6JkwjXOYr0sJraCosfZT2BLMcec/RVERuTumqdFVqIk6j/GoW0t73kvH1FPYoIk+YQLZ7+pd2Vg/jQTq6V666FJWvPkFSDeNVJ3a74YmCkxNAVfdaSEY83VOUIPKiaALAJ5uzOEVNDq/LZO4=

### Context

We have team subscription and can not merge code from second member's branch.

### Visual documentation

[https://github.com/FlutterFlow/flutterflow-issues/assets/78797181/2524fb56-df77-43c7-af14-ffee6b7e4126](https://github.com/FlutterFlow/flutterflow-issues/assets/78797181/2524fb56-df77-43c7-af14-ffee6b7e4126)

![Screenshot 2024-01-11 at 7 26 29 PM](https://uploads.linear.app/3241089d-736a-4422-900b-051e72accdbe/8112a507-1ca9-44e2-a56d-97d113911efc/90cbcc0a-e0a1-48d8-bae4-33aaf11610a3?signature=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXRoIjoiLzMyNDEwODlkLTczNmEtNDQyMi05MDBiLTA1MWU3MmFjY2RiZS84MTEyYTUwNy0xY2E5LTQ0ZTItYTU2ZC05N2QxMTM5MTFlZmMvOTBjYmNjMGEtZTBhMS00OGQ4LWJhZTQtMzNhYWYxMTYxMGEzIiwiaWF0IjoxNzA3MTgyMDI2LCJleHAiOjMzMjc3NzQyMDI2fQ.chcCHgVcRJ8QMz7mRfWNVf8be54Vl8MMu3LCOHIy8n0 "=1438x")

### Additional Info

*No response*

### Environment

```
- FlutterFlow version: v4.0
- Platform: Web
- Browser name and version: Chrome 120.0.6099.199 (Official Build) (x86_64)
- Operating system and version affected: macOS Ventura 13.6
```

**Comment by victoriahuang1** [comment]
Hi @infocodiste , thanks for the report.

> When I merge another branch into the main one, it ends up overwriting all the changes from the other branch, and the main branch loses its modifications.

This is the expected behavior -- when you merge, the changes are applied to both branches (main and the other one).

> Additionally, attempting to rebase with the main branch causes the other branch to adopt the characteristics of the main branch and loses its modifications.

This is also the expected behavior for rebasing. Rebasing will take changes from main and put them onto your other branch.

One thing that you can do is take a look at the project before merge/rebase. You can do so during the merge/rebase by clicking the "Edit Project" button. Previewing and downloading code should also work at this point.
What the project looks like at this stage is what your main / branch will look like afterwards. If you see inconsistencies between here and the what the branches look like after, please provide more precise details on exactly what inconsistencies you observe.

**Comment by infocodiste** [comment]
Hello @victoriahuang1 , thank you for your prompt response and for addressing the issues. I've created a video that showcases the problem I'm encountering during the merge process. When I merge with the main branch, I lose changes from other branch, and the same issue occurs in the other branch as well. I kindly request you to review the shared video and provide guidance on how to proceed. Your insights will be greatly appreciated. Thank you!

Uploading Screen Recording 2024-01-18 at 12.30.51 PM.mov…

**Comment by smian0** [comment]
Hi @infocodiste, we are facing the same issue and thank you for opening a support case. @victoriahuang1, we are eagerly waiting for your guidance. We lost days of work we tried to merge our changes back to main. At this point, we are extremely nervous about doing parallel development due to this issue.

**Comment by victoriahuang1** [comment]
@infocodiste , this could be the expected behavior. It's all dependent on everything that has happening in the project and its branches since their creations. It's hard for me to confirm if this is indeed a bug based on this video.

Can you do a screen recording on a blank project? Starting from the blank and up until you can point out the issues during merge / rebase

**Comment by htphuc1994** [comment]
Hi @infocodiste, I am suffering from this issue as well. When I merge feature branches to the main branch sometimes it loses the new code on both branches. And you know, It affects the team's progress quite much. 
Thank you so much for your investigation.

**Comment by infocodiste** [comment]
@victoriahuang1 I'm sorry to express this concern, but it **seems challenging** to consider this **as expected behavior**. If this were the case, it might discourage individuals from **opting for a team subscription**. The essence of teamwork and branching on Flutterflow lies in the collaborative efforts of team members. After both team members have invested time in completing their respective tasks and are **ready to merge the code**, the **current system implies** potential **loss of work** from one of the branches.

**Comment by smian0** [comment]
Very much in agreement with @infocodiste. Can we escalate this issue so it can be addressed ASAP? @victoriahuang1, what do you need from us to highlight this as a major issue and receive the appropriate attention?

**Comment by steven-ck** [comment]
@victoriahuang1 I experience the same issue. I tried to merge changes to main branch and end up with losing all changes in both target and source branch. It does not always happen and I don't know the exact trigger. I have to create a backup branch for my source branch every time I want to merge/rebase to/onto main. My team is not able to work in collaboration because of this and as a consequence slow down our progress.

**Comment by victoriahuang1** [comment]
In order to look into this more, we need this to be reproduced from a blank project.

**Comment by infocodiste** [comment]
While the demo provided in docs seemed to work well, it's important to note that this isn't always the case. When I initially started working solo, everything was fine. However, when we expanded the team and added a new member, issues arose. After the new team member completed their work on a feature branch, attempts to merge it were unsuccessful from the outset. It's crucial to investigate the root cause of the merge problem. Additionally, I believe it's not necessary for you to provide a demo from the beginning to reproduce the issue, as the problem doesn't seem related to that aspect.

**Comment by MartinLockheart** [comment]
Adding to this thread: My team member and I tried to merge a branch a few days ago with the same results reported above. A lot of progress was almost lost (fortunately, I had backed everything up before running the merge), but until this is fixed, all of the work done on other branches is in limbo. 

I know that I've been able to merge from other branches before and the changes made on both branches (as long as they didn't conflict with each other) were all kept. This seems to be a relatively new issue.

**Comment by MartinLockheart** [comment]
My team successfully merged two branches into our main branch today without any issues, so it seems to be fixed (at least for us).

**Comment by MamadAsap** [comment]
same issue for us... 

We had worked with two builder on a project, so we decided to use branches, but we had the same issue: 

- The first user was making logic changes to one page
- The second user was making UI updates in other pages

The  first user merged his branch on main and closed it. He checked on the main that all his updates were present.
A few hours later, the second user merged his branch onto the main and his changes overrode ALL the changes to the branch that had been merged by the first user...

This was very problematic and caused us to lose several days' work...

**Comment by goodfella2000** [comment]
> My team successfully merged two branches into our main branch today without any issues, so it seems to be fixed (at least for us). 

Thanks for sharing. Simple use case seems to work but when the merge becomes a bit more than a happy path merge, then merge fails along with changes lost in the ether. There is no way to recover the changes. Can you point me to the release notes that addressed this issue as fixed?

**Comment by goodfella2000** [comment]
We would like to get some stats on how many team users are successfully branching and merge changes on a daily basis on average. This will suggest to us if the feature is of sufficient quality so we can try to do branch-based development. Currently, we've suspended all branch based development due to this bug. Can flutterflow team please share stats on the usage of branch-based development?

**Comment by marieloomi** [comment]
Same problem here.

I'm working on a Team plan, so we've created several branches to work in parallel on our project.
Now that I want to merge my branch into the main, when setting up the merge, FlutterFlow tells me there's no conflict, even though the page content is now different between my branch and my main. After that, FlutterFlow copies the contents of my branch to the main, deleting the contents of the main.

We bought the Team plan especially for this feature, as so I'm quite shocked to see that this issue is know since almost a month and nothing is done to fix it, and that FF still allow people to buy this plan.

For the one also having this problem, we were able to get back our work thanks to the snapshot of the branch/main, but still in two different branches. I'm hoping that we will be able to merge it later with the code.

**Comment by victoriahuang1** [comment]
We have not been able to reproduce this. And we need the information requested above in order to investigate further.

Please note that no conflict does not mean no changes. What we want to confirm here is whether the changes displayed during the merge are the same as what you see in the project after the merge. On our end, we have not been able to set it up such that the changes during the merge are not the same as the project after the merge, which is what it seems this ticket is about.

**Comment by marieloomi** [comment]
I tried again today and got a different result than yesterday, still losing some information. You can find the details of the problem in the screen recording attached. I'm showing on my screen the snapshot of my branch before the merge, and the main after the merge. https://github.com/FlutterFlow/flutterflow-issues/assets/158298858/ae4d4494-61a9-43c0-83be-ad7ee8dbf892

I also tried to reproduce it from a blank project but the merge worked correctly.

I can undo my merge with my project snapshot for you to take a look at it, but I will (again) loose time and work. How can we process ? Because I won't continue to work on my projet if I have to go back to an old version for you to see the problem ...

**Comment by smian0** [comment]
Hi @victoriahuang1 and the FlutterFlow team,

I would like to suggest an enhancement to your existing GitHub integration process to improve transparency and control for parallel development workflows. The current UI-driven merge process has led to serious challenges that are difficult to troubleshoot, and it is creating apprehension about potential work loss when dealing with real projects that involve multiple team members. To build greater trust in Flutterflow, I propose the following approach for teams that are utilizing GitHub integration with their own repositories:

Branch Creation: When a branch is generated via the FlutterFlow Branching UI, ensure it's also created directly in the team's GitHub repository.
Direct Links: Incorporate a direct hyperlink within the FlutterFlow Branching UI that navigates the developer to the specific branch on their GitHub repository.
Merge Operations: Execute merge operations directly within the GitHub repository. Developers can see the diffs on their repos. Perhaps the diffs will not be useful due to code-gen but at least it is accessible to developers.
Merge Visibility: Enable developers to directly observe merge operations within GitHub, allowing them to diagnose failures or confirm successful merges on the target branch.
Developer Control: By managing branching and merging within their own GitHub repositories, developers can thoroughly review code changes, perform reverts, initiate recoveries, apply fixes, and more importantly, communicate any merge-related issues back to the FlutterFlow team.
Increased Confidence: Granting developers oversight over branch and merge operations within their own repository mitigates the uncertainty associated with opaque "magic" that Flutterflow Branching UI performs, ensuring reliability and predictability in their development efforts.

I believe these refinements will significantly enhance the developer experience by merging the intuitive FlutterFlow UI with the transparency of branching/merging process. It will also instill greater confidence in the Flutterflow's branching/merging reliability that are key for any serious team development.

Please consider this as high priority request as it will greatly benefit FlutterFlow's goal of attracting larger and more sophisticated enterprise teams. Adopting these changes can make FlutterFlow more appealing to organizations seeking greater control and transparency in their development workflows.

Thank you!

**Comment by victoriahuang1** [comment]
@smian0 thank you for the suggestions. I can put these down as feature requests for the team. We appreciate your patience as we continue working on branching and merging capabilities.

**Comment by victoriahuang1** [comment]
@marieloomi unfortunately, I do not have the information I need from this video. Please see above for the required information needed to debug the specific type of bug. The issue with undoing is that the end result could be different.

Instead, where you could do is start a merge, and instead of directly clicking, complete merge, inspect the project at that state in the middle of the merge. Check to see if there's anything missing and if there is, please call out the specific widget or setting for me to verify, and please leave the project in the middle of the merge until I am able to verify it

**Comment by dantebgarcia** [comment]
Wow. Wish this thread was more prominent on the FF community. I just went through the process of working with a contractor and having them maintain a separate branch to be merged. I've run into a problem myself of not being able to accept a resolution for an updating conflict. But reading this I'll go make a duplicate of my main branch now just in case!

**Comment by PVNDAMVN** [comment]
Same problem here, whenever we try to merge into branch there are always few missing data. When it comes to extreme disaster, it doesn't even show any conflict and does nothing but say successfully merged.
I wish the merge feature will soon be fixed as it is crucial for our work.

**Comment by Twenty45** [comment]
We have the same problem here. It is a bit upsetting to see that this BUG has been reported almost 2 months ago and nothing has been done about it. We can't continue our work this way.

**Comment by JulianMarc1** [comment]
Same problem here. Cant merge branch into main. Flutterfow says my changes have been carried over, but nothing is changed.

**Comment by JulianMarc1** [comment]
Or even worse, sometimes it says there are conflicts in parts of the app that haven been modified. Once it said there were conflicts in the authorization page, but there were no changes there.

**Comment by ready4sea** [comment]
Still waiting as well for resolution, and have been for more than a month. Clearly if the way it works is intented behavior, then the use case is **not adequate**. FlutterFlow, this is another customer paying the Team plan here, and very worried as well.

**Comment by guibertjulien** [comment]
I have same issue. 

1. update develop branch
2. update flutterflow branch
3. pull request from flutterflow branch to develop branch
4. **resolve conflicts** (automatic commit to flutterflow branch)
5. the next pull request, i lost my updates from develop branch

I have issue when i have and resolve conflicts.

We need best practices to manage "code from IDE" and "code from flutterflow". Thx

**Comment by marinosabijan** [comment]
it works good if you use local github merging, not using the online UI on github... for some reason it commits your develops commits to the flutterflow branch which then makes it overwrite/delete your code next time you merge... it doenst happen fi you merge with using local github... the issue is that when using the online UI to merge from flutterflow into develop it also commits your develop into flutterflow - which causes the later issues.

i think github changed the way their UI handles the pull request, making it 2 way... which is bad for this usecase...

**Comment by liviaWee** [comment]
Guys, the merging issue is still happening. Our team just went through that. We have decided to use the branching feature to improve our development process, but ended up losing all the work we had done before. Very upsetting. Is there any feedback to this? 
@victoriahuang1 

I don't know if this is clear for you guys, but this is a huge problem. And you should really get your team to take a look at this... so many people are being affected, that you shouldn't need to rely solo in an evidence to try and solving the problem.

**Comment by goodfella2000** [comment]
Hey all, it's been a while with no fix for "team development" on our FlutterFlow plan, and it's costing us time and data loss. I'm leaning towards asking for a big discount or maybe even a refund unless they fix it quickly. Sometimes hitting them in the wallet gets things moving. Thoughts? Did anyone try this route?

**Comment by syg-eliseu** [comment]
**### PROOF OF THE PROBLEM**

FlutterFlow team, we're taking a hit with the branching feature. Until then I was a lover of this feature until I realized that all the work of a team of 13 devs was being thrown in the trash. I recorded a video of a project being created from Zero.

**[FIRST PROBLEM - DELAY IN SAVING CHANGES AND CREATING BRANCHES]**

I recorded a video of a project being created from Zero.

https://github.com/FlutterFlow/flutterflow-issues/assets/88388127/5bd1ef44-b6e1-47d1-8023-3920ace62766

Steps:

1. Create project

2. Add text widgets to the main branch by placing the time in one of them to serve as a base, wait for it to be saved by checking the snapshot list.

3. After the save is listed in the snapshot list, create a new develop branch.

**[First error found]**
The branch is created from a state prior to Main.

4. Switch branches to check the visual differences in the project and realize that there are differences where there should not be.

5. Rebase 'develop' onto 'main', and notice that in the review the main branch does not visually correspond to what we see in the project.

6. Abort the rebase.

7. Open the main branch, check the list of snapshots and notice that one more appears.

8. Create a new branch from main 'develop 2'

**[Second error found]**
The 'develop 2' branch is also created in a state before the branch.

9. From there, continue creating new branches from main and realize that it will not bring the current state of main.

=====================================================================================

https://github.com/FlutterFlow/flutterflow-issues/assets/88388127/afdd643f-ab44-4b88-81df-29b649ecd4bf


In the second video we can see the continuation of the tests. In addition to the delayed save, we noticed another issue.

[MERGE PERFORMED FALSELY]

Steps:

1. Open the previously created branch from main 'develop 3'. (Remember that it was created incorrectly with an old version of main)

2. Add an image anywhere.

3. Merge with the main branch.

[Third error: False merge]
The project indicates that it was merged but no changes were made.

====================================================================================

https://github.com/FlutterFlow/flutterflow-issues/assets/88388127/de469b52-e56f-4fa8-8f97-3499b8708a13

In the third and final video we finally try to test an attempt to rebase the main branch in develop 3 (Remembering that this branch was created in a previous state)

Steps:

1. Rebase 'develop 3' to main.

2. Open the branching history.
(The branching history does not update so I switch branches to update).

3. Check that a rebased is marked, which in practice is false as it did not bring any changes to the main branch.

==================================================================================

In short, the problems are:

1. Saves are delayed or don't happen.

2. Fake rebases

3. Fake merges

In this simple project we were able to notice these problems that several users have reported.

In a small project it is simple to notice FF's failures, but in a large project it is impossible to notice whether my changes were actually saved or whether my merges and rebases were actually successful.

We urgently need attention on this issue, I'm a Tech Lead at the company I work for and I'm under pressure to keep our 15 projects within FF or migrate to another platform or continue with native development in Flutter without low code.

I hope that with this extremely material evidence the FlutterFlow Team will respond to the complaints. Please make me keep loving FF.

**Comment by leoweebt** [comment]
We're facing this with 4 people working on the project, right now we're all working in the main branch and left the branch feature a side.  This is really bad. Please make increase the severity of this bug.

**Comment by leighajarett** [comment]
Hi everyone. First of all, so sorry you're dealing with this. This is really not okay. We have made this an urgent bug on our side and are actively looking into it. Will report back ASAP

**Comment by leighajarett** [comment]
Just a quick update that we're still working on a solution here. But it's a top priority that's actively being worked on. Thank you so much for your patience!!

**Comment by jollyaccounts** [comment]
Is there any update on this issue? as we are facing same issue on our end on Team Subscription.

**Comment by leighajarett** [comment]
We are actively working on a fix and hope to have it available in production soon!

**Comment by leighajarett** [comment]
THis issue has been resolved. If you run into any merging issues, please submit as a new issue. Thank you so much, and once again - we're really sorry you experienced this issue!

**Comment by asmengistu** [comment]
Hi everyone - Abel here, I’m one of the founders. I just wanted to give you all an apology. This whole thread is difficult to read. I can see how this problem was causing a lot of suffering and unfortunately it wasn’t escalated high enough. I'm just finding out about this and it shows that we have broken processes.

I’m sorry you faced this issue, it shouldn’t have happened, but even worse we should have been more proactive in getting it resolved and showing more empathy. We will do better. Thanks for your patience. 💜
